annotators/user_knowledge_memorizer/server.py

```diff
@@ -0,0 +1,536 @@
+            substr2kind = entity_substr.capitalize()
+            kinds_to_add.append(substr2kind)
+            parents.append(entity_kind)
+            abstract_entity_id = "/".join(["Abstract", substr2kind])
+            abstract_triplets.append((user_id, relationship_kind, abstract_entity_id))
+        else:
+            non_abstract_triplets.append(entity)
+
+    logger.debug(f"abstract_kinds_to_add -- {kinds_to_add}")
+
+    if kinds_to_add:
+        try:
+            graph.ontology.create_entity_kinds(kinds_to_add, parents)
+        except ValueError:
+            logger.info(f"All entity kinds '{kinds_to_add}' are already in KG")
+        except Exception:  # TODO: replace with Terminusdb DatabaseError
+            graph.ontology.create_entity_kinds(parents)
+            try:
+                graph.ontology.create_entity_kinds(kinds_to_add, parents)
+            except ValueError:
+                logger.info(f"All entity kinds '{kinds_to_add}' are already in KG")
+    return abstract_triplets, non_abstract_triplets
+
+
+def check_and_add_properties(graph, prop_triplets: List[dict], user_id: str) -> Tuple[list, list]:
+    """Checks if the property triplets exist in kg and adds them if not."""
+    properties_to_add_to_kg, properties_already_in_kg = [], []
+    try:
+        user_properties = graph.get_properties_of_entity(user_id)
+    except Exception:
+        user_properties = []  # new user
+    for triplet in prop_triplets:
+        if triplet["property"] in user_properties and triplet["object"] == user_properties[triplet["property"]]:
+            properties_already_in_kg.append(triplet)
+        else:
+            if triplet["property"] == "misc attribute":
+                triplet.update({"property_family": set})
+            else:
+                triplet.update({"property_family": Optional})
+            properties_to_add_to_kg.append(triplet)
+
+    if properties_to_add_to_kg:
+        property_kinds = [triplet["property"] for triplet in properties_to_add_to_kg]
+        properties_families = [triplet["property_family"] for triplet in properties_to_add_to_kg]
+        objects = [triplet["object"] for triplet in properties_to_add_to_kg]
+        logger.info(
+            f"property_kinds -- {property_kinds}\nproperties_families -- {properties_families}\n"
+            f"properties_to_add_to_kg -- {properties_to_add_to_kg}"
+        )
+        graph.ontology.create_property_kinds_of_entity_kind(
+            "User",
+            property_kinds,
+            properties_type_families=properties_families,
+        )
+        graph.create_or_update_properties_of_entity(user_id, property_kinds, objects)
+        for prop in properties_to_add_to_kg:
+            del prop["property_family"]
+    return properties_to_add_to_kg, properties_already_in_kg
+
+
+def get_result(request, graph):
+    uttrs = request.json.get("last_human_annotated_utterance", [])
+    utt = uttrs[0]
```

Comment:
WHY? You have a batch of utterances but work only with the first one! you always in all components must consider a batch of inputs. and return a batch of results of the same length as input batch

Reply:
done